static/js/narrow.js

```diff
@@ -182,18 +184,55 @@ exports.activate = function (raw_operators, opts) {
     // Populate the message list if we can apply our filter locally (i.e.
     // with no backend help) and we have the message we want to select.
     if (narrow_state.get_current_filter().can_apply_locally()) {
+        if (select_strategy.flavor === 'first_unread') {
+            var unread_info = narrow_state.get_first_unread_info();
+
+            if (unread_info.flavor === 'found') {
+                select_strategy = {
+                    flavor: 'exact',
+                    msg_id: unread_info.msg_id,
+                };
+            } else if (unread_info.flavor === 'not_found') {
+                if (message_list.all.fetch_status.has_found_newest()) {
+                    select_strategy = {
+                        flavor: 'last_id',
+                    };
+                    activate_status = 'ready_to_preload';
+                }
```

Comment:
I think adding a long comment in here is a good idea, explaining why for `not_found` this logic is correct

Reply:
(Also, should there be an else clause for not `has_found_newest` to set to `pending_server`?